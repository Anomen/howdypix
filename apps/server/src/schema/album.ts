import { objectType } from "nexus";

export const Album = objectType({
  name: "Album",
  definition(t) {
    t.field("name", { type: "String" });
    t.field("dir", { type: "String" });
    t.field("source", { type: "String" });
  }
});
