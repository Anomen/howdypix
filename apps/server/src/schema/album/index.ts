export * from "./albumTypes";
export * from "./albumQuery";
