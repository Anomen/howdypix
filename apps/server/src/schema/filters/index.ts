export * from "./filtersTypes";
export * from "./filtersQuery";
