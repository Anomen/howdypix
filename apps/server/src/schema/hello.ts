import { objectType } from "nexus";

export const Hello = objectType({
  name: "Hello",
  definition(t) {
    t.id("id");
    t.string("toto", { nullable: true });
    t.string("type", { nullable: true });
  }
});
