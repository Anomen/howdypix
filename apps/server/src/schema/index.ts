export * from "./album";
export * from "./auth";
export * from "./user";
