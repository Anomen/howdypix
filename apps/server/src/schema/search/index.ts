export * from "./searchTypes";
export * from "./searchQuery";
