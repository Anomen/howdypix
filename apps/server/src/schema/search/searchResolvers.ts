import {
  NexusGenArgTypes,
  NexusGenFieldTypes,
  NexusGenRootTypes,
} from "@howdypix/graphql-schema/schema.d";
import { appDebug, generateThumbnailUrls } from "@howdypix/utils";
import { SearchResult as EntitySearchResult } from "../../entity/SearchResult";
import { appConfig } from "../../config";
import { ApolloContext } from "../../types.d";
import { searchHelpers } from "../../helpers/searchHelpers";

const { doSearchWithCache } = searchHelpers;

const debug = appDebug("gql");

export const getSearchResolver = () => async (
  root: {},
  args: NexusGenArgTypes["Query"]["getSearch"],
  ctx: ApolloContext
): Promise<NexusGenFieldTypes["Query"]["getSearch"]> => {
  debug(`Fetching album ${args.album}.`);

  const searchResults: EntitySearchResult[] = await doSearchWithCache(
    ctx.connection,
    args
  );

  return {
    photos:
      searchResults?.map(
        (
          searchResult: EntitySearchResult
        ): NexusGenRootTypes["SearchPhoto"] => ({
          id: searchResult.photo.id.toString(),
          thumbnails: generateThumbnailUrls(appConfig.webapp.baseUrl, {
            file: searchResult.photo.file,
            dir: searchResult.photo.dir,
            source: searchResult.photo.source,
          }).map((tn) => tn.url),
          file: searchResult.photo.file,
          birthtime: Math.round(searchResult.photo.birthtime),
        })
      ) ?? [],
  };
};
