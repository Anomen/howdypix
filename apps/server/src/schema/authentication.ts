import { enumType, mutationField, objectType, stringArg } from "nexus";
import { createTransport } from "nodemailer";
import smtpTransport from "nodemailer-smtp-transport";
import { state } from "../state";
import { find } from "lodash";
import { NexusGenFieldTypes } from "@howdypix/graphql-schema/schema";
import { magickLink } from "../email";

export const SendEmailMessage = enumType({
  name: "SendEmailMessage",
  members: ["AUTH_EMAIL_OK", "AUTH_EMAIL_ERR_NOT_EXIST", "AUTH_EMAIL_ERR"],
  description:
    "The type of message that the user can get when requesting a magic link."
});

export const SendEmailType = objectType({
  name: "SendEmailType",
  definition(t) {
    t.field("messageId", { type: "SendEmailMessage" });
  }
});

export const sendEmail = mutationField("sendEmail", {
  type: "SendEmailType",
  args: {
    email: stringArg()
  },
  resolve: async (root, args) =>
    new Promise<NexusGenFieldTypes["Mutation"]["sendEmail"]>(resolve => {
      const user = find(state.userConfig.users, u => u.email == args.email);
      console.log(user);

      if (user) {
        const transporter = createTransport(
          smtpTransport({
            host: process.env.SMTP_HOST,
            port: parseInt(process.env.SMTP_PORT as string),
            ignoreTLS: true
          })
        );

        transporter.sendMail(
          {
            from: `${state.userConfig.emailSender.name}<${state.userConfig.emailSender.email}>`,
            to: `${user.name}<${user.email}>`,
            html: magickLink({ code: "123", name: user.name })
          },
          (error, info) => {
            if (error) {
              resolve({
                messageId: "AUTH_EMAIL_ERR"
              });
            } else {
              resolve({
                messageId: "AUTH_EMAIL_OK"
              });
            }
          }
        );
      } else {
        resolve({
          messageId: "AUTH_EMAIL_ERR_NOT_EXIST"
        });
      }
    })
});
