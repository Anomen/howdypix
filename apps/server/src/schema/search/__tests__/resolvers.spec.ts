import { Connection } from "typeorm";
import { NexusGenFieldTypes } from "@howdypix/graphql-schema/schema.d";
import * as searchResolversModule from "../searchResolvers";
import {
  doSearch,
  fetchSearchResult,
  findSavedSearch,
  saveNewSearch,
  saveSearchResult
} from "../searchHelpers";
import { SearchResult as EntitySearchResult } from "../../../entity/SearchResult";
import { Search as EntitySearch } from "../../../entity/Search";

jest.mock("../searchHelpers", () => ({
  findSavedSearch: jest.fn(),
  saveNewSearch: jest.fn(),
  saveSearchResult: jest.fn(),
  doSearch: jest.fn(),
  fetchSearchResult: jest.fn()
}));

describe("getSearchResolver", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  const callResolver = async (): Promise<NexusGenFieldTypes["Query"]["getSearch"]> => {
    const connection = ({
      getRepository: jest.fn(() => ({}))
    } as unknown) as Connection;
    const resolver = searchResolversModule.getSearchResolver();
    return resolver({}, {}, { user: null, connection });
  };

  describe("with not-saved results", () => {
    beforeEach(async () => {
      (findSavedSearch as jest.Mock).mockReturnValue(null);
      (saveSearchResult as jest.Mock).mockReturnValue([
        {
          order: 0,
          id: 123,
          search: {
            id: 123
          },
          photo: {
            id: 123,
            file: "file",
            source: "source",
            dir: "dir"
          }
        }
      ] as EntitySearchResult[]);
    });

    test("should save the new search", async () => {
      await callResolver();
      expect(saveNewSearch).toHaveBeenCalled();
    });

    test("should make the search in the library", async () => {
      await callResolver();
      expect(doSearch).toHaveBeenCalled();
    });

    test("should save the search results", async () => {
      await callResolver();
      expect(saveSearchResult).toHaveBeenCalled();
    });

    test("should return the results", async () => {
      expect(await callResolver()).toMatchSnapshot();
    });
  });

  describe("with saved results", () => {
    beforeEach(() => {
      (findSavedSearch as jest.Mock).mockReturnValue({
        searchResults: [],
        filterBy: "",
        orderBy: "orderBy",
        source: "source",
        album: "album",
        id: 123
      } as EntitySearch);

      (fetchSearchResult as jest.Mock).mockReturnValue([
        {
          order: 0,
          id: 123,
          search: {
            id: 123
          },
          photo: {
            id: 123,
            file: "file",
            source: "source",
            dir: "dir"
          }
        }
      ] as EntitySearchResult[]);
    });

    test("should NOT save a new search or do a new search", async () => {
      await callResolver();
      expect(saveNewSearch).not.toHaveBeenCalled();
      expect(doSearch).not.toHaveBeenCalled();
      expect(saveSearchResult).not.toHaveBeenCalled();
    });

    test("should fetch the results from the saved search", async () => {
      await callResolver();
      expect(fetchSearchResult).toHaveBeenCalled();
    });

    test("should return the saved photos", async () => {
      expect(await callResolver()).toMatchSnapshot();
    });
  });

  /*
  test("should make a new search if not in the search table", async () => {
    const saveSearch = jest.fn();
    const saveSearchResult = jest.fn();

    const connection = ({
      getRepository: (
        entity: Function
      ):
        | Partial<Repository<EntitySearch | EntitySearchResult | EntityPhoto>>
        | undefined => {
        if (entity.name === "Search") {
          return {
            findOne: jest.fn(),
            save: saveSearch
          };
        }

        if (entity.name === "SearchResult") {
          return {
            find: jest.fn(),
            save: saveSearchResult
          };
        }

        if (entity.name === "Photo") {
          return {
            find: jest.fn().mockReturnValue([
              {
                id: 1234
              },
              {
                id: 567
              },
              {
                id: 890
              }
            ] as EntityPhoto[])
          };
        }

        return undefined;
      }
    } as unknown) as Connection;

    await resolver({}, {}, { user: null, connection });
    expect(saveSearch).toHaveBeenCalled();
    expect(saveSearchResult).toHaveBeenCalledTimes(3);
  });
  */
});
