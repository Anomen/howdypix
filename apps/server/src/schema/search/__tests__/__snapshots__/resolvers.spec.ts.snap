// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSearchResolver with not-saved results should return the results 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": NaN,
      "file": "file",
      "id": "123",
      "thumbnails": Array [
        "http://localhost:3004/static/@source:dir/filex200.jpg",
        "http://localhost:3004/static/@source:dir/filex600.jpg",
        "http://localhost:3004/static/@source:dir/filex1280.jpg",
        "http://localhost:3004/static/@source:dir/filex2048.jpg",
      ],
    },
  ],
}
`;

exports[`getSearchResolver with saved results should return the saved photos 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": NaN,
      "file": "file",
      "id": "123",
      "thumbnails": Array [
        "http://localhost:3004/static/@source:dir/filex200.jpg",
        "http://localhost:3004/static/@source:dir/filex600.jpg",
        "http://localhost:3004/static/@source:dir/filex1280.jpg",
        "http://localhost:3004/static/@source:dir/filex2048.jpg",
      ],
    },
  ],
}
`;
